File: src/main/java/ch/obermuhlner/astro/gradient/filter/PseudoMedianFilter.java
package ch.obermuhlner.astro.gradient.filter;

import ch.obermuhlner.astro.image.ArrayDoubleImage;
import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.color.ColorModel;

public class PseudoMedianFilter implements Filter {

  private final int radius;
  private final ColorModel model;

  private final HorizontalMedianFilter horizontalMedianFilter;
  private final VerticalMedianFilter verticalMedianFilter;

  public PseudoMedianFilter(int radius, ColorModel model) {
    this.radius = radius;
    this.model = model;
    horizontalMedianFilter = new HorizontalMedianFilter(radius, model);
    verticalMedianFilter = new VerticalMedianFilter(radius, model);
  }

  @Override
  public DoubleImage filter(DoubleImage source, DoubleImage target, int width, int height) {
    ArrayDoubleImage temp = new ArrayDoubleImage(width, height, target.getColorModel());
    horizontalMedianFilter.filter(source, temp, width, height);
    verticalMedianFilter.filter(temp, target, width, height);
    return target;
  }

  @Override
  public String toString() {
    return "PseudoMedianFilter(radius=" + radius + ")";
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/filter/PseudoMedianFilter.kt
package ch.obermuhlner.astro.gradient.filter

import ch.obermuhlner.astro.image.ArrayDoubleImage
import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel

class PseudoMedianFilter constructor(private val radius: Int, private val model: ColorModel) : Filter {
    private val horizontalMedianFilter: HorizontalMedianFilter = HorizontalMedianFilter(radius, model)
    private val verticalMedianFilter: VerticalMedianFilter = VerticalMedianFilter(radius, model)

    override fun filter(source: DoubleImage, target: DoubleImage, width: Int, height: Int): DoubleImage {
        val temp = ArrayDoubleImage(width, height, target.colorModel)
        horizontalMedianFilter.filter(source, temp, width, height)
        verticalMedianFilter.filter(temp, target, width, height)
        return target
    }

    override fun toString(): String {
        return "PseudoMedianFilter(radius=$radius)"
    }

}